### PR TITLE
Fix broken background effects in preview (except song name and fireworks)

### PR DESCRIPTION
**Pull request by bencodeorg**
Fixes preview for the following background effects:

- Diamonds
- Snow
- Space
- Stars
- Laser Dance Floor

Companion PR in code-dot-org: https://github.com/code-dot-org/code-dot-org/pull/54456

https://github.com/code-dot-org/dance-party/assets/25372625/d90dee7e-bbfc-4a54-a5b0-976d92d4b9d1

## Links

- trello: https://trello.com/c/9RbgDFNW/74-update-dance-party-background-effects-which-do-not-show-up-on-initial-preview-load

## Testing Story

I tested manually (in the main workspace / visualization column) that each background I changed functioned as expected in preview and in a normal run (see video above).

I did not test every background in the modal, but I did confirm that at least basic rendering of the preview continued to function when using the AI modal.

## Follow-up work

I'm going to fix the song name background separately from this PR, as the work will be a little different (ie, we need to make it so that the `dance-party` repo can always access the currently selected song metadata). Sanchit may also need access to song metadata for the live preview work he's looking into, so I'll coordinate with him on an approach that works for both of us.

### CODE REVIEW

File: src/Effects.js [hunk 1]
```diff
@@ -1063,23 +1069,24 @@ module.exports = class Effects {
...
-          let y = 200 * p5.sin(p5.frameCount);
-          if (y < 0) {
-            y *= -1;
+        for (const [index, laser] of this.laser.entries()) {
+          let x = this.getPreviewCustomizations().getX(index);
+          if (x < 0) {
+            x *= -1;
           }
-          const angle = p5.atan2(laser.y, y);
+          const angle = p5.atan2(laser.y, x);
```
Comment (bencodeorg) [inline src/Effects.js:1089]: According to p5 docs, second argument to `atan2` is x, so I renamed. If that doesn't make sense to ppl who remember trig better than I do, let me know :)

https://p5js.org/reference/#/p5/atan2

Comment (bencodeorg) [reply src/Effects.js:1088]: Also, the w, x, and z parameters are unused, so I removed them.

File: src/Effects.js [hunk 2]
```diff
@@ -1558,6 +1598,17 @@ module.exports = class Effects {
       },
 
       draw: function () {
+        // Note: the fireworks effect is really an overlay of many draws.
```
[the remark this answers is not in the document]
Comment (bencodeorg) [reply src/Effects.js:1600]: Curious for your thoughts here @breville

[the remark this answers is not in the document]
Comment (breville) [reply src/Effects.js:1600]: This is a tricky one.  

A delay of 100ms on a fast device does feel worrisome, as one could imagine this turning into a significant fraction of a second on a lower-powered device, which would be a noticeable delay.  The other ingredient is how often, and when, a preview is rendered; for example, is a student going to encounter UI hitches as they are trying to manipulate blocks?  

Would an alternate strategy be to pre-generate a PNG with the result of 250 calls, save that in our repo, and then simply render that when we want to preview the effect?  We'd need to include opacity for each pixel, of course.

[the remark this answers is not in the document]
Comment (bencodeorg) [reply src/Effects.js:1600]: we'd be calling preview any time a student makes a change to their code, so yeah if this is closer to a second on slower devices, not great.

a png is an interesting thought! will give that a go.

[the remark this answers is not in the document]
Comment (bencodeorg) [reply src/Effects.js]: @breville @sanchitmalhotra126 per discussion offline, pulling out the fireworks preview effect for now so we can get the rest of these in.

File: src/Effects.js (hunk 1 above)
Comment (fisher-alice) [reply src/Effects.js:1088]: So much math!

**Pull request merged**